Mathlib/CategoryTheory/MorphismProperty/Composition.lean

```diff
@@ -154,30 +138,87 @@ end naturalityProperty
 
 /-- A morphism property is multiplicative if it contains identities and is stable by
 composition. -/
-class IsMultiplicative (W : MorphismProperty C) extends W.ContainsIdentities : Prop :=
-  stableUnderComposition : W.StableUnderComposition
-
-lemma comp_mem (W : MorphismProperty C) {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z) (hf : W f) (hg : W g)
-    [IsMultiplicative W] : W (f ≫ g) :=
-  IsMultiplicative.stableUnderComposition f g hf hg
+class IsMultiplicative (W : MorphismProperty C)
+    extends W.ContainsIdentities, W.IsStableUnderComposition : Prop :=
 
 namespace IsMultiplicative
 
 instance op (W : MorphismProperty C) [IsMultiplicative W] : IsMultiplicative W.op where
-  stableUnderComposition := fun _ _ _ f g hf hg => W.comp_mem g.unop f.unop hg hf
+  comp_mem f g hf hg := W.comp_mem g.unop f.unop hg hf
 
 instance unop (W : MorphismProperty Cᵒᵖ) [IsMultiplicative W] : IsMultiplicative W.unop where
-  id_mem' _ := W.id_mem _
-  stableUnderComposition := fun _ _ _ f g hf hg => W.comp_mem g.op f.op hg hf
+  id_mem _ := W.id_mem _
+  comp_mem f g hf hg := W.comp_mem g.op f.op hg hf
 
 lemma of_op (W : MorphismProperty C) [IsMultiplicative W.op] : IsMultiplicative W :=
   (inferInstance : IsMultiplicative W.op.unop)
 
 lemma of_unop (W : MorphismProperty Cᵒᵖ) [IsMultiplicative W.unop] : IsMultiplicative W :=
   (inferInstance : IsMultiplicative W.unop.op)
 
+instance : (isomorphisms C).IsMultiplicative where
+  id_mem _ := isomorphisms.infer_property _
+  comp_mem f g hf hg := by
+    rw [isomorphisms.iff] at hf hg ⊢
+    infer_instance
+
+instance : (monomorphisms C).IsMultiplicative where
+  id_mem _ := monomorphisms.infer_property _
+  comp_mem f g hf hg := by
+    rw [monomorphisms.iff] at hf hg ⊢
+    apply mono_comp
+
+instance : (epimorphisms C).IsMultiplicative where
+  id_mem _ := epimorphisms.infer_property _
+  comp_mem f g hf hg := by
+    rw [epimorphisms.iff] at hf hg ⊢
+    apply epi_comp
+
+instance {P : MorphismProperty D} [P.IsMultiplicative] (F : C ⥤ D) :
+    (P.inverseImage F).IsMultiplicative where
+
 end IsMultiplicative
 
+/-- A class of morphisms `W` has the two-out-of-three property if whenever two out
+of three maps in `f`, `g`, `f ≫ g` are in `W`, then the third map is also in `W`. -/
+class HasTwoOutOfThreeProperty (W : MorphismProperty C)
+    extends W.IsStableUnderComposition : Prop where
+  of_postcomp {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z) : W g → W (f ≫ g) → W f
+  of_precomp {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z) : W f → W (f ≫ g) → W g
```

Comment:
I wonder if we should define these classes separately. There are plenty classes that only satisfy one of them: mono, epi, finite scheme morphisms etc.

But maybe there isn't much API to develop and there isn't much gain.

Reply:
Mono and epi satisfy an even more stronger conditions: if the composition is mono, then the first map is. Indeed, I do not think this would be a significant gain.